pkg/storage/etcd/etcd_watcher.go

```diff
@@ -136,37 +138,59 @@ func newEtcdWatcher(list bool, include includeFunc, filter storage.FilterFunc, e
 
 // etcdWatch calls etcd's Watch function, and handles any errors. Meant to be called
 // as a goroutine.
-func (w *etcdWatcher) etcdWatch(client *etcd.Client, key string, resourceVersion uint64) {
+func (w *etcdWatcher) etcdWatch(ctx context.Context, client etcd.KeysAPI, key string, resourceVersion uint64) {
 	defer util.HandleCrash()
 	defer close(w.etcdError)
+	defer close(w.etcdIncoming)
 	if resourceVersion == 0 {
-		latest, err := etcdGetInitialWatchState(client, key, w.list, w.etcdIncoming)
+		latest, err := etcdGetInitialWatchState(ctx, client, key, w.list, w.etcdIncoming)
 		if err != nil {
 			w.etcdError <- err
 			return
 		}
-		resourceVersion = latest + 1
+		resourceVersion = latest
 	}
-	_, err := client.Watch(key, resourceVersion, w.list, w.etcdIncoming, w.etcdStop)
-	if err != nil && err != etcd.ErrWatchStoppedByUser {
+
+	opts := etcd.WatcherOptions{
+		Recursive:  w.list,
+		AfterIndex: resourceVersion,
+	}
+	watcher := client.Watcher(key, &opts)
+	w.ctx, w.cancel = context.WithCancel(ctx)
+
+	for {
+		resp, err := watcher.Next(w.ctx)
+		if resp != nil {
```

Comment:
This is slightly strange.
Can we do it like this:

if err != nil {
  w.etcdError <- err
  return
}
w.etcdIncoming <- resp

I think this is clearer